src/main/kotlin/org/elm/ide/test/core/LabelUtils.kt

```diff
@@ -50,10 +50,10 @@ object LabelUtils {
         return String.format("%s://%s", protocol, pathString(path))
     }
 
-    fun fromLocationUrlPath(path: String): Pair<String, String> {
+    fun fromLocationUrlPath(path: String, testsRelativeDirPath: String): Pair<String, String> {
```

Comment:
I'm not sure if you've explored this, but what if the "label" fully encoded the path to the Elm file so that we don't need to reconstruct it later?

For instance, in `org.elm.ide.test.core.ElmTestJsonProcessor#newTestSuiteStartedEvent` when we construct the location URL, we could embed the full, relative path to the Elm file instead of just the name of the module. Then you wouldn't need to make any changes in `ElmTestLocator` *and* it would make it easier in the future to handle multiple test roots.

Can you give that a try to see if it makes things simpler?

Reply:
Funnily enough that was the first route I went down. The problem I found was that the path to a test is defined as a URL such as `elmTestTest://TheModule/some/other/stuff`.  Here, `TheModule` is the path to the test file (in this case "TheModule.elm"). If the file is in a subfolder the URL would be `elmTestTest://TheFolder.TheModule/some/other/stuff`. The important point here is that the name of the module can be extracted from the URL by looking in the first part of the path, as seen in `LabelUtils.fromLocationUrlPath`.  This module name is used, for example, when showing the test results in the treeview in IntelliJ.

Now, say the URL was expanded to include the full path to the file, such as this: `elmTestTest://src/test/frontend/TheFolder.TheModule/some/other/stuff`: the code wouldn't know exactly which part contained the module name.  So instead I think we'd have two options:
1. Store some parts(s) of the data in `metainfo` (one of the parameters received in `ElmTestLocation.getLocation`, which we can set, for example, in `ElmTestJsonProcess.newTestSuiteStatedEvent`: we just need to call a different constructor of `TestSuiteStartedEvent`, which receives the `metainfo`).  We'd have to do the same thing for `Test` as for `TestSuite`.
2. Come up with some sort of custom syntax that would allow us to extract the name of the module from the full file path. I've just noticed that a similar approach is already taken in `LabelUtils.kt` in the `ErrorLabelLocation` class, where it puts various data separated by `::`.

It was at this point of my investigations that I decided the other approach I eventually went with would be simpler, especially as I wanted to keep the PR smaller and more self-contained.

So, I totally agree with your suggested approach, and like I say it's the first thing I thought of too, but I wonder if it merits own separate PR?  What do you think?  And do you have a preference between the two options above, or any other ideas?

Reply:
I would prefer front-loading it, either by 
- encoding it in the label URL (using some kind of separator, as you suggested).
- or by changing the URL to be the full path to the file on disk and using the `metainfo` parameter for the Elm module name.

My preference would be for the `metainfo` option, but the first option might be less disruptive. The existing `LabelUtils` code is pretty gnarly, so it might be a pain either way.

In terms of how to go about it, I would keep this PR open and just create a new one to see how that comes out.

Reply:
All sounds good to me. I'll give it a bash and submit a PR when it's ready. Thanks!

Reply:
Spent a bit of time digging, and it seems to me that the least intrusive way of making this change would be to keep the URL as-is, and to store the path to the root of the `ElmProject` in the `metainfo`. That way we can combine the `metainfo` with the first part of the URL to get to the full path to the actual Elm file. Any other approach would require a larger-scale change I think, especially in the unit tests.

The JavaDocs about `metainfo` in `SMTestLocator` say this:

> The <code>metainfo</code> parameter simplifies the search for locations, but can not be used to identify the test. A good example for <code>metainfo</code> is the line number of the beginning of the test. It can speed up the search procedure, but it changes when editing.

So it's kinda debatable as to whether storing the location of the project in `metainfo` would be an advisable use it. I think you could probably argue it either way.

You happy for the PR to go down this route, for the sake of reducing the number of changes?  Or would you prefer a larger-scale refactor of this area of the code?

Reply:
Go with the simplest change. I'm not worried about "abusing" the `metainfo` parameter.

I didn't write the elm-test integration, and it has been my intention to rewrite it to match the way the rest of the plugin is written, but I just haven't had a chance to do it yet.

Reply:
#595 now implemented and merged into master, so I think this review comment can be resolved?  This PR has now been updated with all changes from master, including the changes in #595.